packages/apollo-language-server/src/utilities/graphql.ts

```diff
@@ -87,31 +93,180 @@ export function removeDirectives(ast: ASTNode, directiveNames: string[]) {
   });
 }
 
-// remove fields where a given directive is found
-export function removeDirectiveAnnotatedFields(
-  ast: ASTNode,
+/**
+ * Recursively remove orphaned fragment definitions that have their names included in
+ * `fragmentNamesEligibleForRemoval`
+ *
+ * We expclitily require the fragments to be listed in `fragmentNamesEligibleForRemoval` so we only strip
+ * fragments that were orphaned by an operation, not fragments that started as oprhans
+ */
+function removeOrphanedFragmentDefinitions<AST extends ASTNode>(
+  ast: AST,
+  fragmentNamesEligibleForRemoval: Set<string>
+): AST {
+  /**
+   * Flag to keep track of removing any fragments
+   */
+  let anyFragmentsRemoved = false;
+
+  // Aquire names of all fragment spreads
+  const fragmentSpreadNodeNames = new Set<string>();
+  visit(ast, {
+    FragmentSpread(node) {
+      fragmentSpreadNodeNames.add(node.name.value);
+    }
+  });
+
+  // Strip unused fragment definitions. Flag if we've removed any so we know if we need to continue
+  // recursively checking.
+  ast = visit(ast, {
+    FragmentDefinition: {
+      enter(node) {
```

Comment:
Why are you using `enter` here and `leave` below?

Reply:
I intended to remove all `FragmentDefinition` that are flagged for removal and then remove all associated spreads.

My understanding is that all the nodes would be visited with `enter` before _any_ would be visited by `leave`. Is that correct?

Reply:
Ah, not that isn't quite how it works.

```
{
  me {
    name
  }
  bar
}
```

Would be `enter(me), enter(name), leave(me), enter(bar)`

Reply:
Ah! So this should be split into operations to accomplish what I was after. 👍

Reply:
This makes less sense than I'd originally thought; there's no need to look at `FragmentSpread`s here because we're removing anything that isn't in the list generated above, which includes all `FragmentSpread`s.